src/SFML/Graphics/Shape.cpp

```diff
@@ -284,49 +306,87 @@ void Shape::updateTexCoords()
+            twiceArea += m_vertices[i + 1].position.cross(m_vertices[i + 2].position);
+        return twiceArea >= 0.f;
+    }();
 
+    // In each iteration, d1 and n1 vectors will be the direction and normal of the segment before the current one.
+    auto [d1, n1] = computeDirectionAndNormal(m_vertices[count].position, m_vertices[1].position, flipNormals);
+
+    std::size_t outlineIndex = 0;
     for (std::size_t i = 0; i < count; ++i)
     {
         const std::size_t index = i + 1;
 
-        // Get the two segments shared by the current point
-        const Vector2f p0 = (i == 0) ? m_vertices[count].position : m_vertices[index - 1].position;
+        // Get the segment starting at the current point
         const Vector2f p1 = m_vertices[index].position;
         const Vector2f p2 = m_vertices[index + 1].position;
 
-        // Compute their normal
-        Vector2f n1 = computeNormal(p0, p1);
-        Vector2f n2 = computeNormal(p1, p2);
-
-        // Make sure that the normals point towards the outside of the shape
-        // (this depends on the order in which the points were defined)
-        if (n1.dot(m_vertices[0].position - p1) > 0)
-            n1 = -n1;
-        if (n2.dot(m_vertices[0].position - p1) > 0)
-            n2 = -n2;
-
-        // Combine them to get the extrusion direction
-        const float    factor = 1.f + (n1.x * n2.x + n1.y * n2.y);
-        const Vector2f normal = (n1 + n2) / factor;
-
-        // Update the outline points
-        m_outlineVertices[i * 2 + 0].position = p1;
-        m_outlineVertices[i * 2 + 1].position = p1 + normal * m_outlineThickness;
+        // Compute its direction and normal pointing towards the outside of the shape
+        const auto [d2, n2] = computeDirectionAndNormal(p1, p2, flipNormals);
+
+        // Decide whether to add a bevel or not
+        const float twoCos2            = 1.f + n1.dot(n2);
+        const float squaredLengthRatio = m_miterLimit * m_miterLimit * twoCos2 / 2.f;
+        const bool  isConvexCorner     = 0.f <= d1.dot(n2) * m_outlineThickness;
+        const bool  needsBevel         = twoCos2 == 0.f || (squaredLengthRatio < 1.f && isConvexCorner);
+
+        if (needsBevel)
+        {
+            // Make room for two more vertices
+            m_outlineVertices.resize(m_outlineVertices.getVertexCount() + 2);
+
+            // Combine normals to get bevel edge's direction and normal vector pointing towards the outside of the shape
+            const float    twoSin2   = 1.f - n1.dot(n2);
+            const float    sin       = std::sqrt(twoSin2 / 2.f);
+            const Vector2f direction = (n2 - n1) / twoSin2; // Length is 1 / sin
+            const Vector2f extrusion = (flipNormals != (0.f <= d1.dot(n2)) ? direction : -direction).perpendicular();
+
+            // Compute bevel corner position in (direction, extrusion) coordinates
+            const float u = m_miterLimit * sin;
+            const float v = 1.f - std::sqrt(squaredLengthRatio);
```

Comment:
```suggestion
            // Combine normals to get bevel edge's direction and normal vector pointing towards the outside of the shape
            const float    twoSin2   = 1.f - n1.dot(n2);
            const Vector2f direction = (n2 - n1) / twoSin2; // Length is 1 / sin
            const Vector2f extrusion = (flipNormals != (0.f <= d1.dot(n2)) ? direction : -direction).perpendicular();

            // Compute bevel corner position in (direction, extrusion) coordinates
            const float sin = std::sqrt(twoSin2 / 2.f);
            const float u   = m_miterLimit * sin;
            const float v   = 1.f - std::sqrt(squaredLengthRatio);
```
Lets move `sin` farther down closer to where it's used.

Reply:
It is "used" in a comment but ok.